### ISSUE TITLE
"deprecation" property

### ISSUE THREAD

**Issue by awwright** [json-schema-org/json-schema-spec#74]
Create a schema keyword that specifies an instance (property in an instance, usually) as deprecated -- it shouldn't be used, but exists only for internal use, or for reverse compatibility with obsolete or old products.

**Comment by handrews** [comment]
+1 to this.  The point here, I assume, is for documentation and perhaps to allow implementations to issue warnings, rather than to affect validation.

**Comment by Relequestual** [comment]
Interesting proposal.

I know that JSON Schema isn't specifically for API's, but we should consider how API's might version their endpoints. Could this proposal actually be part of a seperate doc about API driven information. For example, included in deprecated, how it's deprecated and if it's replaced, and possibly other upgrade path type data.
Very much a straw man comment, not well thought through.

**Comment by handrews** [comment]
@Relequestual I'm not sure how far we want to chase the versioning idea here, but my approach to API evolution (rather than versioning) is to only change the version of the schema, and retain compatibility by (as much as possible) avoiding both `required` and `"additionalProperties": false`.  I use the following rules:

* Resources add new schema versions to their list of profiles, and do not remove old versions
* If an instance validates against a schema that is a valid profile for the resource, no matter what version that schema is, than the instance can be used according to that version of the schema.
    * So if your code was written for a specific version, as long as that still validates, you can keep using that without worrying about the new version.
    * This can allow for continuing to function in a scenario where some new instances may not validate against the old schema, but if the instances you work with still validate, you don't have to update to support the things you don't need.  For instance, a value was added to an enum, but you never use the new value.
* If you can't keep compatibility at all, make a new resource and handle the change through HATEOAS, use of 410 GONE, redirects, etc.
* There are no versions in the URI or primary media type (the schema versions likely show up in their profile URIs, since those need to be distinct somehow)

**Comment by Relequestual** [comment]
Interesting. Most of the time our API (and others I've worked on) have needed "new versions" because of new fields or new allowed values. Sometimes that's backwards compatible, but often it isn't for me, as much as I'd like it to be.

**Comment by handrews** [comment]
As long as you avoid setting additionalProperties to false, then new fields should be fine (unless you're generating strictly typed code off of the schemas without any provision for unrecognized fields, but I avoid that in favor of more generic client libraries).

Enumerations get tricky, but that's where the "if it still validates you can still use it" rule comes into play- technically the new version is not fully compatible but most instances will validate against either version.

My approach to API evolution is to do as much as possible to allow users to keep using the API as they have been unless *their specific usage* is broken by the update.  It requires some design discipline, and the willingness to add a new resource as a replacement and deprecate the old instead of doing a major version change on the entire API and breaking a bunch of clients that don't actually care one way or the other.

**Comment by handrews** [comment]
Admittedly, I haven't actually had a chance to prove that this can work long-term, but it's something I'm trying with the current project I'm working on.

**Comment by Relequestual** [comment]
I think Protocol Buffers allows for flagging deprecation. If someone has time, may be worth seeing what that system allows for.

**Comment by awwright** [comment]
#173 is a sufficient solution for me, so I'll close this out in favor of the PR

**Comment by handrews** [comment]
@awwright please don't close issues while the corresponding PR is still open.  PRs are not always accepted.

**Comment by awwright** [comment]
@handrews A PR is just a type of issue on GitHub, and I can indefinitely re-edit it if necessary, or re-open this issue of mine, or someone can file a new issue for me if it really comes down to that

**Comment by handrews** [comment]
@awwright This is yet another instance of you randomly choosing how to use GitHub in whatever ways that suit you, no matter whether anyone else finds it confusing.  Some issues you want to keep open, some you don't, we just get to guess at what you'll complain about, and if we want to keep track of things in some consistent way, we're out of luck if you decide not to.

All of the discussion about wanting a clear and defined process that the community has had clearly means nothing to you.  Your response to any complaint or request for clarity is always just "well I see it this way so that's what I'm doing".

**Comment by awwright** [comment]
@handrews It's my issue! Anyone is allowed to close their own issue for any reason. There's nothing special about it.

**Comment by Relequestual** [comment]
I'll re-open this issue as I agree issues that do not have accepted pull requests should not be closed.

**Comment by Relequestual** [comment]
It seems like there's quite a lot of discussion and decisions to be made on this issue. Specifcially the comments in the PR review: https://github.com/json-schema-org/json-schema-spec/pull/173#discussion_r91867794

It's clear there are a number of unresolved abiguities.

> In a root schema with a "self" link is means that the resource itself is deprecated? But it's still there currently?
> 
> In a root schema without a "self" link it means....?
> 
> In a property subschema (in any of the property keywords) it means that all matching properties are deprecated but will still be present?
> 
> In a schema applying to a single position in an array/tuple ("items" is an array) it means...?
> 
> In a schema applying to multiple array positions ("items" as one schema, or "additionalItems") it means...?
> 
> In the schema for "schema" or "targetSchema" it means...?
> 
> This concept is not ready for a PR. There are far too many cases that need discussing and a coherent view of hypermedia resources that need to be developed before we can just toss this in.

**Comment by Relequestual** [comment]
I've set this as priority: low. People aren't crying out for it, and it doesn't fix an existing bug. It would be nice to go in the next release if the above issues can be resolved, but otherwise it shouldn't become a blocker.

**Comment by discordier** [comment]
I just ran into the same wall. I want to denote in a schema that a property will get removed in a later version.
So currently I have a schema with id: `https://example.org/schemas/1.0/payload-schema.json`.
Now we are releasing the next version adding new properties and types named `https://example.org/schemas/1.1/payload-schema.json` but also want to deprecate old stuff which got superseeded by the new types to be ultimately get removed in `https://example.org/schemas/2.0/payload-schema.json` (long term future).

Simply as all document types evolve over time you need a migration phase (1.x schema series in above example).
So we want all 1.x documents to validate against all 1.x schemas. This is possible, as pointed out by @handrews above when "avoiding both `required` and `"additionalProperties": false`, all types are preserved and only non mandatory new stuff gets added (See also SemVer).
Now we also want validation to break when the deprecated stuff is validated against the 2.x schema.

All this is currently possible from validation POV but we lack a way to tell the user of the schema (developer) about the migration phase. Of course this can be handled by defining proprietary keys, used by whatever validator one is using or attaching a new validator to validate twice, but this will cause fragmentation of standards.
Therefore defining this right in the spec would be better - or as @Relequestual stated "nice to have".

Yet, I wonder how big need is as next to no one seems to feel the need (with me being the second guy in this thread apparently).
The fact that XML XSDs also lack a standardized way to define something like this makes me wonder if the schema really is the right place to add this but I fail to see a better alternative (which can be utilized by CI and yet not break production).

**Comment by handrews** [comment]
@discordier we've generally had trouble finding anyone who has even used hyper-schema enough to participate in discussions.  So don't take the low count as *necessarily* indicating lack of support.  I encourage you to comment on any hyper-schema issues and pull requests you see- it is a challenge for us to decide on directions when only two or three people have opinions they are comfortable expressing.  Here are the issues and PRs (including but not limited to hyper-schema) that we're targeting for the next draft: https://github.com/json-schema-org/json-schema-spec/milestone/2

@Relequestual marked this low priority because we're trying to quickly resolve enough things to publish Draft 6 and this is both not essential for that immediate effort and because there are several questions without obvious answers.  I'd personally be in favor of getting this into Draft 7, though, if we can't sort it out in time for Draft 6.

I'm only familiar with XSD in the most basic sense, but I gather it's more analogous to JSON Schema Validation rather than JSON Hyper-Schema.  I definitely think this feature belongs in hyper-schema assuming we can sort out the details.

**Comment by lidio601** [comment]
+1

**Comment by Relequestual** [comment]
@lidio601 Please avoid +1 comments. You can *thumbs up* any comment you agree with. Thanks.

**Comment by brendajin** [comment]
I would also like deprecation schema, or some type of meta-tagging for fields that are being added, replaced, and deprecated. As others mentioned before, it should have a date to accompany it.

**Comment by philsturgeon** [comment]
Howdy! Where did this conversation end up? I know deprecations are something that Open API v3.0 has a little of, and they're [considering improvements for v3.1](https://github.com/OAI/OpenAPI-Specification/issues/782). 

It would be amazing to see some collaboration on this issue, so we can help reduce the now _very narrow_ gap between JSON Schema and Open API v3.x.

**Comment by handrews** [comment]
@philsturgeon everyone here has been kind of caught up in other things, but there are signs of life and we do hope to have another draft out before the current ones expire in October.

The deprecation idea is popular so I hope it makes it into the next draft.

The last PR stalled because there were many ambiguities and we did not have the time and focus to resolve them.  The use cases I see (which may not be what anyone else agrees with) are:

* "deprecated" in a Link Description Object (LDO) deprecates the link (this could mean that the target resource is deprecated, or just that the connection between the resources will be removed).

* "deprecated" in an LDO with "rel": "self" indicates that the resource is deprecated

* "deprecated" inside of an object field schema (values within "properties" or "patternProperties", or the value of "additionalProperties") means that the relevant properties are deprecated, but the containing object is not.

The semantics of deprecation and array elements are unclear, so I'd avoid that on the first pass.  If use cases emerge we can work through them.

Deprecating a schema that is not an object field schema has no clear semantics to me.  In my opinion, only LDOs and fields schemas are reasonable use cases for deprecation (although I may be missing some obvious other ones).

**Comment by awwright** [comment]
We may be running into an issue similar to "required" where it was decided that it doesn't make sense as a boolean, but only as a list of properties on an object, so it was turned into an array listing key names that are required.

Maybe the same thing is true of "deprecated", but I'd also be interested in providing a rationale as to why the property is deprecated, maybe at least a link to a document.

I vaguely remember a proposal to add a link relation doing exactly this, pointing to a document that explains how the origin resource is deprecated, but I can't find any evidence of such a thing, maybe I invented that.

**Comment by philsturgeon** [comment]
Absolutely. So this is exactly the same conversation we're having over on @dret's [Sunset](https://tools.ietf.org/html/draft-wilde-sunset-header-02) header draft. 

Outstanding questions of where to send humans and computers are [being discussed here](https://github.com/dret/I-D/issues/60).

Sunset will take a date, and optionally a URL to some metadata, and thats about it. 

Some people like versions, some people like dates, but for Sunset it's going date only at this point. Versions are too app-specific for a header, but JSON Schema could probably have some opinions on version perhaps. Not sure. 

Either way, theres some crossover here.

**Comment by dret** [comment]
just to clarify: `Sunset` as currently defined (feedback very welcome) *only* takes a date and nothing else:

```
Sunset = HTTP-date
```

https://github.com/dret/I-D/issues/60#issuecomment-319739804 mentions the possibility of linking to additional information using a link relation, but that's not part of the current draft.

**Comment by handrews** [comment]
@dret @philsturgeon I'd prefer a date over a version because it's more flexible.  Some APIs use evolution techniques rather than versions, and that's my preferred approach.  I'll have to catch up on the `Sunset` discussion.  I like having link relations for things but I haven't read the rationales for or against yet.

@awwright Agreed on this probably making more sense as a list parallel to `required`.  "What about this object is deprecated?" would be a lot easier to reason about than "is this schema part of a thing that can be deprecated and if so is it?"

It looks like the `Sunset` header and an LDO-level `deprecated` would work together, while a field-level `deprecated` would be a different thing.

**Comment by dret** [comment]
On 2017-08-03 09:25, Henry Andrews wrote:
> @dret <https://github.com/dret> @philsturgeon 
> @awwright <https://github.com/awwright> Agreed on this probably making 
> more sense as a list parallel to |required|. "What about this object is 
> deprecated?" would be a lot easier to reason about than "is this schema 
> part of a thing that can be deprecated and if so is it?"

i am confused by this whole discussion. to me it looks like this:

- if i make things mandatory in my structure, that's a promise for these 
structures to be there all the time. i have to keep the promise.

- if i don't want to make this promise, then i make things optional and 
this makes it clear that things may not always be there.

- the proposed mechanism seems to aim at adding a facility that allows 
me to say when i am planning to break a promise that i previously made? 
what's the exact benefit of this, and how is it any different from 
making things optional from the start and therefore requiring clients to 
be able to handle the presence or absence of a structure?

**Comment by awwright** [comment]
@dret It sounds like we may have to differentiate how this is used for (1) defining expected user input, and (2) annotating the output of a JSON API. What "deprecated" means in these scenarios may be slightly different.

I've been understanding it as an API value that may not be provided in the future, and that you shouldn't use it, or should adjust your application to rely on some other value.

**Comment by dret** [comment]
On 2017-08-03 13:39, Austin Wright wrote:
> @dret <https://github.com/dret> It sounds like we may have to 
> differentiate how this is used for (1) defining expected user input, and 
> (2) annotating the output of a JSON API. What "deprecated" means in 
> these scenarios may be slightly different.

and there are many different ways in which these differences can 
manifest, which makes it hard to capture them in a predefined model.

> I've been understanding it as an API value that may not be provided in 
> the future, and that you shouldn't use it, or should adjust your 
> application to rely on some other value.

so what makes it different from something that's optional? other than 
you might start labeling it "optional" not from the very start, but at 
some point in time during the lifetime of your API?

**Comment by handrews** [comment]
@dret I prefer to avoid declaring anything as required and build my tooling appropriately.  However, declaring fields to be required and then removing them later is an extremely common real-world occurrence, no matter how I feel about it as a design choice.

**Comment by dret** [comment]
On 2017-08-03 17:04, Henry Andrews wrote:
> @dret <https://github.com/dret> I prefer to avoid declaring anything as 
> required and build my tooling appropriately. However, declaring fields 
> to be required and then removing them later is an extremely common 
> real-world occurrence, no matter how I feel about it as a design choice.

interesting. if you build things your way, there's no need to 
communicate deprecation: everything can go away anytime anyway, and 
clients have to be prepared for that. afaict, that makes it hard to 
build applications that are based on some minimal core contract, but if 
that's something that works for you, it seems that you are all set!

**Comment by handrews** [comment]
> if that's something that works for you, it seems that you are all set!

In practice, I doubt I can manage it all the time, but it's something I am going to explore and see how much I can get away with :-)

Ideally, the only reason to drop a field is that it is legitimately no longer relevant.  My preference is that optional fields are always absent when not relevant, never present with a `null` value.  If clients are built to facilitate working with fields that may be absent, then a certain category of "removed" fields are no longer a problem.

Obviously some fields can't work that way.  If you're configuring something related to DNS, you need the domain name.  Otherwise nothing makes sense at all.  But there are other things that are a lot less critical.  Hopefully over the next six months to a year I'll be able to get some real-world validation (or invalidation) of various ideas in this area.

**Comment by adamvoss** [comment]
Is this hyper-schema specific?  Or rather, is it out of scope for JSON Schema?

I know I have encountered at least one case where configuration settings have changed with Visual Studio Code (which provides a schema for its configuration file) where they potentially could have captured the change in a deprecation message rather than leaving users wondering why the file no longer validated.

**Comment by handrews** [comment]
This, `readOnly`, and the `media` object all at least theoretically are usable outside of Hyper-Schema.  `media` could arguably be viewed as extended validation (in the same way as `format`, in that it conveys information that could be but need not necessarily be validated in full).

Alternatively, Hyper-Schema is not limited to APIs.  This is one of the key points that confuse people about Hyper-Schema.  It is a hypermedia format, not an API description format.  So just like people often use hypermedia formats like HTML (or even more commonly, Markdown) to store local documents, there's no reason you couldn't use a hyper-schema with Visual Studio configuration settings.

**Comment by adamvoss** [comment]
> there's no reason you couldn't use a hyper-schema with Visual Studio configuration settings.

I must be missing something.  Where do the semantics come from?

For example, if I am validating a config (or other) file against a schema and the file uses a deprecated property I think the validator SHOULD emit a warning about the deprecated property.  If "deprecated" is only part of hyper-schema, how would a JSON Schema validator know to issue a warning?

**Comment by handrews** [comment]
You would obviously need a hyper-schema-aware implementation, or at least one that makes it easy to register the keywords you care about as extension keywords.

Hopefully with draft-06 and beyond, hyper-schema implementations will become more common.  My work in that area got kind of put on hold for a while but I am getting back to it.  But the most popular validators tend to allow extension keywords.

We also haven't specified what an implementation SHOULD do, and would probably leave it fairly open.  Some systems may want it more as a documentation thing, others may want runtime behavior.  This is similar to `default`, which is why many implementations ignore it by, um.. default :-) but provide options to write missing defaults into the instance during validation.  Implementations can always layer functionality on top of the spec.

**Comment by adamvoss** [comment]
I think that is a valid view.  Though, I would throw in a word of caution about extensions and fragmentation, in that one of the advantages of JSON Schema as I see it is that (on the whole) it is well-defined and compatible.   Something that is implemented as extensions (and then ones that are loosely defined with respect to behavior), makes me hesitant to use it unless I _really_ need the functionality because the number of implementations supporting my schema suddenly drop and the number that support it but no not behave how I intend may be non-zero.

In response to my original question though, it sounds like it is considered out of scope for JSON Schema.

**Comment by dret** [comment]
> In practice, I doubt I can manage it all the time, but it's something I 
> am going to explore and see how much I can get away with :-)

keep us posted.

> Ideally, the only reason to drop a field is that it is legitimately no 
> longer relevant. My preference is that optional fields are always absent 
> when not relevant, never present with a |null| value. If clients are 
> built to facilitate working with fields that may be absent, then a 
> certain category of "removed" fields are no longer a problem.

agreed in principle. in practice, it is very hard to build anything 
meaningful when you can count on nothing. my guidance is almost opposite 
to yours: design and define a stable core and commit to it, allowing 
clients to rely on that core.

http://dret.typepad.com/dretblog/2016/04/robust-extensibility.html

> Obviously some fields can't work that way. If you're configuring 
> something related to DNS, you need the domain name. Otherwise nothing 
> makes sense at all. But there are other things that are a lot less 
> critical. Hopefully over the next six months to a year I'll be able to 
> get some real-world validation (or invalidation) of various ideas in 
> this area.

like i said, keep us posted. the art of defining and evolving APIs is 
definitely something that will only become more important as we see 
things becoming increasingly connected. how to do it well and robustly 
is important, and a space where we need more experience and guidance.

**Comment by handrews** [comment]
OK, trying to push this towards a new concrete proposal:

Going from @awwright's observations that this is more like `"required"` than anything else, I see two options for the object field case:

* Use an array: `"deprecated": ["foo", "bar"]`, which is equivalent to booleans in individual schemas
* Use an object `"deprecated": {"foo": "2017-09-01", "bar": "2018-01-01"}`, which is equivalent to date-valued fields in individual schemas

As with #117 (`"patternRequired"`) we would likely need `"patternDeprecated"` (and `"additionalDeprecated"`?).  See also #364 (applying this object-level approach to `"readOnly"`).

I slightly lean towards the object form despite probably needing more keywords, as it then parallels the `Sunset` header approach of using dates (assuming @dret is still going in that direction).

For the link/resource case, I'd like to try to handle that through #296 (link usage hints / target attributes) which could then piggy-back on the `Sunset` work (assuming @dret thinks that is valid).  While there would not be a `Sunset` header specified for all protocols / URI schemes, I think there is a general question to be answered in #296 about re-using useful hints outside of their original specification.

I feel like we should be able to make a decision on the object field approach and get that to PR stage, and either handle the link/resource case in #296 or start a new issue to track it on its own.

As for exactly how this is specified for required vs optional fields, I think that will be more easily discussed given a PR with actual wording.  We don't entirely nail down what required and non-required fields actually mean (is required a permanent commitment? it depends on the implementation, and is outside of JSON Schema's scope).

Thoughts?

**Comment by handrews** [comment]
For such a simple concept, this turns out to be difficult to get right (two PRs by two different contributors have now been retracted).

@adamvoss brought up the excellent idea of a `deprecationMessage` functionality of some sort.

Looking at the OpenAPI proposals that @philsturgeon noted, while a lot of their operation-based approach doesn't translate well to hyper-schema, there are other useful concepts like indicating a replacement.

So I wanted to bring this back for a bit more open discussion.

I will comment again soon with specific ideas, although anyone should feel free to jump in on this of course.

**Comment by philsturgeon** [comment]
When I was talking to @dret about the Sunset header (for entire-endpoint deprecations), we came to the agreement that probably a message was not entirely necessary. 

Instead a link is provided via Links with a rel=sunset, and that is either a URL to the new endpoint to use, or a link to some documentation or a blog. Whatever that link is, it will be a lot more useful than whatever was jammed into the reason, because for entire endpoints.... well do you care? 

Human written messages like "We love our customers very much, and decided to listen to their feedback on some of the..." are no good, and the other end of the spectrum is just "Bugs and app updates", neither of which is useful.

Instead the link approach allows clients to build a simple formatted message: 

```
"Endpoint #{env.url} is deprecated for removal on #{datetime.iso8601}. Read more at #{sunset_link}"
```

This may not be the way to go for JSON Schema. I see JSON Schema and GraphQL as having a lot in common, and they do offer a reason: 

https://github.com/graphql/graphql-js/pull/384

This adding of a reason here could easily outline for humans that `foo` is deprecated and the reason is just: "Bar is better because X".  

I feel like #391 is headed in the right direction, but would adding `deprecated` and `deprecatedMessage` get too noisy? 

If not, can we go a little further and add a `deprecatedUntil` or something? The dates to me really are more important than the human words.

**Comment by handrews** [comment]
@philsturgeon I like the link relation for per-resource deprecation.  Assuming PR #383 `"targetHints"` is accepted, that would then look something like:

```JSON
{
    "rel": "some=old-thing",
    "href": "path/to/old/thing",
    "targetHints": {
        "sunset": ["Sat, 31 Dec 2018 23:59:59 GMT"],
        "link": [
            ["/uri/path/to/sunset/resource", {"rel": "sunset"}]
        ]
    }
}
```

**Comment by philsturgeon** [comment]
Maybe I confused things by bringing up sunset, im not sure how targetHints work and dunno if I'd want to overload the term Sunset with stuff. 

Are targetHints designed to hint towards potential response headers, or are the request header orientated? I see in the PR they talk about Allow (req) so these resp headers seem out of place in this form.

**Comment by handrews** [comment]
@philsturgeon maybe read the targetHints PR?  It includes other examples :-)

But briefly:  `"targetHints"` are basically JSON-serialized response headers.  Like `"targetSchema"` and `"mediaType"` they are non-authoritative.  Schema authors can provide them as an optimization (and in the case of HTTP, as a convenience to avoid having to figure out whether something is discoverable over HEAD vs OPTIONS).

So that example just means "if you did a HEAD on this, you would very likely get a `Sunset` header with this date/time, and a `Link` header with this URI and relation type."

**Comment by handrews** [comment]
Request headers are described with schemas instead of values, and are covered in the aptly-named `headerSchema` PR #390

**Comment by handrews** [comment]
@philsturgeon 
> I see in the PR they talk about Allow (req) so these resp headers seem out of place in this form.

Unless I am *really* confused `Allow` is a response header?  Admittedly, I did confuse `Accept` and `Content-Type` the other day, so it's possible that I'm really confused...

**Comment by philsturgeon** [comment]
> Unless I am really confused Allow is a response header? 

Gah sorry brain frazzled. I did the same thing and read Allow as Accept and got messed up. Ok, let me do my homework on targetHints and see if it lines up with the right sorta thing. 

targetHints may well solve endpoint deprecations (letting you know its likley before you hit it and find out) but regular old JSON Schema still needs to have deprecation, message and date to let folks know a field is about to vanish too. Right?

**Comment by handrews** [comment]
> Gah sorry brain frazzled.
well that clearly makes two of us today!  :-D

> regular old JSON Schema still needs to have deprecation, message and date to let folks know a field is about to vanish too. Right?

yes, and that's the main focus of this issue and both of the PRs that attempted to address it.  We'd brought up resource-level deprecation as well, but certainly for draft-07, I'd like to give `"targetHints"` a chance to solve as many problems as it can with the generic mechanism, and only design specific approaches when that fails.  There's a very larger set of things that can be conveyed in response headers and arguing over each individually hasn't really gotten us far.

Anyways, yes, field deprecation needs a totally separate mechanism.

**Comment by handrews** [comment]
As for ideas on per-field deprecation, I see a few options:

1.  Decide it's complicated enough to call for a sub-object.  This would allow for things like clearly distinguishing between date vs version-based deprecation, and providing information about replacements in a structured way:

```JSON
{
  "properties": {
    "foo": true,
    "bar": true
  },
  "patternProperties": {
    "^biz": true
  },
  "deprecated": {
    "foo": {"date": "2017-12-01", "message": "...", replacedBy: "??not sure what to put here"}
  },
  "patternDeprecated": {
    "^biz.*stuff$": {"version": "2.0", "reason": "..."}
  }
}
```

2.  Use a tuple, where the first item is the boolean-or-string field described in PR #391 and the optional second field is the message.  This avoids nested keywords (which we generally avoid) but still allows additional information:

```JSON
{
  "properties": {
    "foo": true,
    "bar": true
  },
  "patternProperties": {
    "^biz": true
  },
  "deprecated": {
    "foo": ["2017-12-01", "message of some sort"]
    "bar": ["2018-01-01"]
  },
  "patternDeprecated": {
    "^biz.*stuff$": ["2.0", "message for this one"]
  }
}
```

3.  Do we want to reconsider a sub-object in the context of all of the annotation fields?  Should there be an annotation structure?  I've just barely started asking this question so no specifics yet.

**Comment by levbishop** [comment]
The #173 proposal, where `deprecated` applies to a sub-schema made sense to me and I don't understand the objections. The obvious interpretation is that any matches to a sub-schema with `"deprecated": true` should be interpreted as "this JSON file is valid according to the current schema, but will no longer be valid in a future revision of the schema". This allows deprecation due to changes due *any* restriction that is expressable in json-schema. In the most complex cases (eg, changes to `dependencies`, changes to regexes in `patternProperties`) it might be necessary to do something like:
```JSON
"oneOf": [ { <new_schema> }, {"deprecated": true, <old_schema>}]
```
or if <new_schema> and <old_schema> aren't easily made mutually exclusive:
```JSON
"oneOf": [ {<new_schema}, {"deprecated": true, "not": {<new_schema>}, <old_schema>}
```
It would allow clearer schemas (and probably it's no extra burden on implementations) to allow use of `anyOf` here, with a rule to suppress deprecation warnings if there are both deprecated and non-deprecated matching branches.

In either case, once the deprecation period is over and the old syntax is dropped, it's a simple update to the schema of deleting the branches with `"deprecated": true` and removing any `anyOf` that may have been rendered superfluous.

Common changes, such as disallowing a previously-allowed field can be done very simply, by adding `"deprecated": true` to the relevant place, without any need for `oneOf`/`anyOf`.

Other deprecation-related fields `deprecationDate`, `deprecationMessage`, `deprecationURL`, etc if decided necessary can live at the same level as `deprecated` and the scope is obvious.

Unlike the #173 proposal, a #391 approach seems limited and ill-defined. For example what is the equivalent of:
```JSON
{
  "anyOf": [ 
    { "type": "integer", "minimum": 0}, 
    { "deprecated": true, "type": "number"}
  ]
}
```

**Comment by dlax** [comment]
Just to mention that the discussion is getting confusing because, as per the title of this issue, we should be talking about hyper-schema; but #391 touches the validation part, hence addressing the "per-field deprecation". (On the other hand, #173 (closed) concerned hyper-schema; but I agree with @handrews that the `targetHints` should be tried first.)

**Comment by handrews** [comment]
@dlax this is an annotation keyword like `readOnly`, so when it was filed (nearly a year ago), Hyper-Schema was the correct place for it.  Hyper-Schema was in the title for historical reasons that are now irrelevant- you still see this (or Validation, or Core) as a prefix on some other issues that I haven't gotten around to re-titling.  Anyway, I changed the title.  Even #173 was really focused more on fields than resources, although it was actually very hard to figure out how it worked in various situations which is why it was rejected.

**Comment by handrews** [comment]
@levbishop thanks for the feedback and use cases.

There's a lot going on not just with this specific proposal but around how annotation keywords (which have no impact on validation, but instead just provide information about usage) should work.

Stepping back even from that, there is a principle that JSON Schema keywords can be interpreted within a single schema object in isolation.  It is not necessary to know whether the schema object is a root schema or subschema.  If it is a subschema, it is not necessary to know whether it is a property schema, an array item schema, or a schema describing some aspect of a hyperlink.

However, deprecation is not a stand-alone concept.  Deprecation occurs within some context.  A specific property member is deprecated within an object.  A resource is deprecated within a hypermedia system.  I'm posting the following questions for a third time, noting that _no one has ever answered them_.  If you want to advocate for `deprecated` as a boolean within a property's schema, you need to answer these questions without violating JSON Schema's architectural principles:

> In a root schema with a "self" link it means that the resource itself is deprecated? But it's still there currently?
> 
> In a root schema without a "self" link, or in a non-Hyper-Schema schema, it means....?
> 
> In a property subschema (in any of the property keywords) it means that all matching properties are deprecated but will still be present? _[this is the one that is well understood, of course]_
>
> In a schema applying to a single position in an array/tuple ("items" is an array) it means...?
>
> In a schema applying to multiple array positions ("items" as one schema, or "additionalItems") it means...?
>
> In the schema for "submissionSchema", "targetSchema", "hrefSchema", "headerSchema",  it means...?

In terms of implementation, how can I tell if a property is deprecated?  Looking in the immediate schema for `properties` is not sufficient.  I need to look down through any `*Of` keywords that are present.  Your `oneOf` example shows this, but consider a very complex schema with multiple levels of `*Of` keywords.  It gets more and more difficult.  This, on its own, is not a fatal problem: figuring out if a link applies to an instance has the same problem, and there is a section in Hyper-Schema addressing the more complex cases.  But it's a factor.  And (getting back to my "larger questions about annotation keywords" comment), the same issue has caused quite a bit of confusion over what to do with `title`, `description`, and `default` in such combinatoric cases.

Applying deprecation at the object schema level, where the keyword values indicate which fields are deprecated in the context of that object, puts the keyword at the level where its effects are felt.  When looking at an object, I want to know which fields are deprecated.  Implementations, for the most part, look at one schema at a time.  If I am looking at a single property's schema, in many implementations I am no longer aware that I'm looking at a nested bit of data.  This simplifies implementation considerably which is why it is such a desirable property.

This is also why there is a proposal to switch `readOnly` to a similar approach.  It's why `required`, which was originally a boolean within a subschema, moved out to the object schema level in draft-04.

**Comment by handrews** [comment]
@levbishop for your example (which I've wrapped in the object-level schema as that's obviously necessary to show the alternative)

```JSON
{
  "type": "object",
  "properties": {
    "foo": {
      "anyOf": [ 
        { "type": "integer", "minimum": 0}, 
        { "deprecated": true, "type": "number"}
      ]
    }
  }
}
```

the equivalent would be something like this (the `"then": true` isn't strictly necessary, but without it one would be forgiven for wondering if the `"then"` got left off.  But it's valid to have `"if"` and `"else"` with no `"then"`).

```JSON
{
  "type": "object",
  "if": {
    "properties": {
      "foo": {"type": "integer", "minimum": 0}
    }
  },
  "then": true,
  "else": {
    "deprecated": {"foo": "2017-10-01"},
    "properties": {
      "foo": {"type": "number"}
    }
  }
}
```

**Comment by handrews** [comment]
At the moment, this feels unlikely to be resolved for draft-07.  If someone can reconcile all of the above concerns and get to a successful PR, I'll happily include it.  But I'm going to put it in the "future" milestone for now.

**Comment by philsturgeon** [comment]
So long as we're 100% focusing on field deprecation in this issue and not hypermedia stuff (that'll be another issue/PR), I'm confident we can get this thing back on track.

It seems to me now, despite my earlier interest in "deprecated" being a keyword in the property's schema, that approach is a non-starter. The [examples from @handrews](https://github.com/json-schema-org/json-schema-spec/issues/74#issuecomment-328365726) make that clear to me. 

GraphQL Types can get away with that simple annotation next to the field declaration as they don't have any of the if/then/else stuff JSON Schema does. 

Is there any agreement on these statements:

1. The `deprecated` keyword will need to work very much like `required`

2. Trying to get messages involved is overkill, as the field is going away and the why doesn't really matter

3. Trying to get versions in when JSON Schema has no opinion on versions is going to be a mess

4. Dates are necessary to avoid panic attacks that a field might go away tomorrow, or to avoid complacency and the warning being ignores. A date provides the right level of urgency for the resolution, even though the date might not be 100% abided to on the server

@Relequestual @handrews @dlax

**Comment by handrews** [comment]
> even though the date might not be 100% abided to on the server

I think this bit from the [Sunset header proposal](https://tools.ietf.org/html/draft-wilde-sunset-header-02) may be relevant

>    Even though the Sunset header information is made available by the
   resource itself, there is no guarantee that the resource indeed will
   become unavailable, and if so, how the response will look like for
   requests made after that timestamp. 

I'd also say there's no guarantee that the field (in our case) won't disappear sooner either, although that would certainly annoy clients/customers and be, um... ill-advised.

**Comment by dret** [comment]
On 2017-09-19 13:10, Henry Andrews wrote:
> I'd also say there's no guarantee that the field (in our case) won't 
> disappear sooner either, although that would certainly annoy 
> clients/customers and be, um... ill-advised.

i think that's the case for all these link hints. it's the same for 
links being annotated with "this is where to find the english version" 
or "this is where to find the PDF version". you'll never know for sure 
until you actually try at runtime. clients need to be prepared for this 
(i.e., shouldn't crash and burn), so making this as explicit as possible 
helps the robustness of the landscape.

**Comment by handrews** [comment]
@dret agreed.  In the rewrite I'm doing for #377 (just reworking for clarity, no implementation-impacting changes), I plan to put a section about this in the **Target Hints** section.  Part of that reorganization is to facilitate making this patterns clear rather than just tacking them onto certain individual fields.

**Comment by philsturgeon** [comment]
@dret just so everyone is all clear, this issue is not about link hints, only field deprecations. Let's make sure from now on we're all only talking about field deprecation. :)

**Comment by handrews** [comment]
@philsturgeon @dret sorry that I was my fault for bringing in the Sunset header.  These things are very separate but analogous in my head so I tend to reference both without thinking that it might be confusing.

**Comment by Relequestual** [comment]
In agreement with @philsturgeon on all counts for this one.
We should make sure the date is in an unambigious format!

**Comment by adamvoss** [comment]
> Trying to get messages involved is overkill, as the field is going away and the why doesn't really matter

It is important to know why in order to know how to update a document to remove the deprecated member.

I think this issue may be running into of intention stemming from different usages of JSON Schema.  Many of the vocal users here seem to be ones who just JSON Schema in APIs and I have even seen comments (not in this item) that JSON is always used as part of a larger system.  This is not true in my usage where I will use manually maintained JSON in-lieu of a database or spreadsheet for storing information (often to be displayed by Jekyll but always) and I will have a schema for it.  Similarly, many tools/applications have user-edited JSON configuration.

Regarding **deprecation** specifically, Microsoft even added a custom extension for JSON Schema to capture a deprecation message.  See: https://github.com/Microsoft/vscode-json-languageservice/blob/a89e995877162b698e338e8301403a4ea49d5d92/src/test/parser.test.ts#L1155-L1171

**Comment by handrews** [comment]
@adamvoss this is going in the validation spec rather than the hyper-schema spec specifically because it applies to more than just APIs/hypermedia.

**Comment by adamvoss** [comment]
I'm confused by your response. I was only trying to say a deprecation
message is important.

**Comment by handrews** [comment]
> I think this issue may be running into of intention stemming from different usages of JSON Schema. Many of the vocal users here seem to be ones who just JSON Schema in APIs and I have even seen comments (not in this item) that JSON is always used as part of a larger system.

I was replying to this part.

I don't have an opinion on the messages at the moment.

**Comment by philsturgeon** [comment]
@adamvoss fair enough on this: 

> It is important to know why in order to know how to update a document to remove the deprecated member.

That's a valid point. Would you be ok with deprecationHref or deprecationInfo or something, which would contain a URL to a place where there is more information? It could be a JSON Schema reference which lets folks know they can replace it, or a link to human readable documentation, or a blog post, etc. 

This is how sunset is avoiding the need for messages, and maybe we could do the same.

**Comment by handrews** [comment]
@philsturgeon @adamvoss if people want to include URIs they should just do so using Hyper-Schema keywords (even if the system is not otherwise hypermedia)

```JSON
{
    "deprecated": ["foo"]
    "properties": {"foo": true, "bar": true}
    "links": [{"rel": "deprecation", "href": "https://example.com/deprecations/foo"}]
}
```

or something like that.  We already have a generic syntax for adding links so I'd rather not just toss in more keywords for specific links.

**Comment by philsturgeon** [comment]
LGTM. So we dont need message or link. Keeping it easy.

**Comment by handrews** [comment]
Also, more reasons to use and implement Hyper-Schema :-)
For this use case, you wouldn't even need a full implementation, but it would hopefully encourage people to think of what else you could do with hyper-schema.

With this approach, a schema author could choose to link to a single resource that explains all deprecations for the object, or individual resources for each deprecated field, whichever makes more sense for the resource's users.

**Comment by dret** [comment]
On Sep 21, 2017, at 15:10, Henry Andrews <notifications@github.com> wrote:
> With this approach, a schema author could choose to link to a single resource that explains all deprecations for the object, or individual resources for each deprecated field, whichever makes more sense for the resource's users.
> 
and if you're *really* good, you'd have a single resource supporting frag ids, so that the individual links still would provide meaningful access to the specific information.

**Comment by adamvoss** [comment]
> > It is important to know why in order to know how to update a document to remove the deprecated > member.
> 
> That's a valid point. Would you be ok with deprecationHref or deprecationInfo or something, which would contain a URL to a place where there is more information? It could be a JSON Schema reference which lets folks know they can replace it, or a link to human readable documentation, or a blog post, etc.

I don't think this would allow for a very good editor experience and it creates an additional burden on a schema creator to have a place to host such information.  Most of the schema at http://schemastore.org/json/ are for things that would be manually edited by a human rather than tooling and many of the schema there are not hosted by the projects that describe them (where requiring a url/hosting would either more coordination from the target project or create a problem for external schema authors).

**Comment by handrews** [comment]
@adamvoss it's possible for the link to be an internal one that simply points to a different place within the schema (since any implementation may add extension keywords, and `contentMediaType` would allow storing, for instance, HTML).  That allows for a great deal of flexibility in proving arbitrary additional deprecation information without needing to include every possibility directly in the spec.

**Comment by handrews** [comment]
I'm not dead-set on the link approach, but I think it is a good extension mechanism, and we will no doubt continue to find areas where some bit of data is important for a specific use case, but not broadly important enough to be worth complicating the primary specification.

It would be good to have an approach for handling these long-tail features.

**Comment by adamvoss** [comment]
The process you specified would only be useful to an editor if it were standardized.  Otherwise the link leads allows for essentially infinite variability which cannot be codified into an algorithm.  Even then, it sounds like a complicated and not very discoverable process.

This post is too long for me to just mention him directly, but if you have a concise specific wonder about the usage, it could be valuable to mentions `aeschli` who might be able to comment on MS's decision to add a custom property.  Here is the issue that implemented that feature: https://github.com/Microsoft/vscode/issues/14260

**Comment by handrews** [comment]
@adamvoss if we're making up the link relation we can make up requirements for its target.

But I'm going to step back and let you and @philsturgeon sort this out while I focus on the getting hyper-schema over the finish line.

**Comment by handrews** [comment]
Earlier, in response to @levbishop, I said:

> In terms of implementation, how can I tell if a property is deprecated? Looking in the immediate schema for properties is not sufficient. I need to look down through any *Of keywords that are present. Your oneOf example shows this, but consider a very complex schema with multiple levels of *Of keywords. It gets more and more difficult. This, on its own, is not a fatal problem: figuring out if a link applies to an instance has the same problem, and there is a section in Hyper-Schema addressing the more complex cases. But it's a factor. 

Having thought about this in several contexts a lot more, I no longer think it's a factor, as there is inevitably a need for this sort of search to support annotation keywords in general.  See #423 for my full thoughts on the matter.  I'm not sure any of it solves the disagreements here, but for people who like consistent underlying theoretical frameworks, well... it's a thing :-P

**Comment by philsturgeon** [comment]
Because some men want to watch the world burn, here's how PayPal handle this with their own vendor prefixed deprecated OpenAPI extensions: https://github.com/paypal/api-standards/blob/master/api-style-guide.md#annotation-x-deprecated

I'd like anyone interested in this deprecated functionality to have a bit of a think about these keywords, and see if they think that the current suggestions fit. 

We don't need to copy PayPal, and they're a company very focused on "API Version" as a planned and published thing, many are not that. They also get into the nitty gritty of deprecating enum values, and I'm not sure I want to live in that world - even though I've 100% had to do that in the past. 

Let's all have a little think on these usages. If others could post with thoughts that'd be cool.

**Comment by handrews** [comment]
Thanks, @philsturgeon, I'll take a look at that.

Also, I invite you, @adamvoss, @levbishop and anyone else interested to take a look at PR #424 which lays out a way of approaching annotations and extended vocabularies, which will include whatever we end up doing for field deprecation.  It might help us all think about how this should work a little more thoroughly.  This is the PR resulting from issue #423 that I mentioned above.  It's probably actually easier to just read the PR and ignore the issue :-P

**Comment by handrews** [comment]
We've done a lot of work defining what annotations are and how they get collected and used during the later parts of draft-07 and into the main work of draft-08.  Which has also made some underlying JSON Schema principles more clear to me.  With that in mind, I would now support adopting the single boolean form for a number of reasons:

1. See #364 for why I'm dropping some other proposals to use arrays instead of booleans.
1.  OpenAPI has a substantial user base setting a precedent.  I have no intention of adopting everything OpenAPI adopts (e.g. `nullable` and `discriminator` do not fit in JSON Schema for a number of reasons), but this was an option under serious consideration as a proposal anyway, and the one @awwright favored as the person who filed the issue.  And it seems nice to keep compatibility if we can.
1.  JSON Schema in general favors annotations that can be interpreted broadly, and allows applications to narrow that interpretation.  Having a simple boolean that can be backed up by application-specific behavior or extensions seems to fit that philosophy.
1. JSON Schema in general prefers to allow harmless nonsense (and if applications come up with a use, great!) rather than trying to prevent keyword combinations that seem useless.  This is relevant further down.
1.  If we really need other data (dates, links, etc.) we can add those as more keywords later.
1.  I have answers for the previously unanswered questions:

> In a root schema with a "self" link is means that the resource itself is deprecated? But it's still there currently?
> In a root schema without a "self" link it means....?
> In the schema for ~~"schema"~~ "submissionSchema" or "targetSchema" it means...?

In all cases it means that the entire representation is somehow deprecated.  Applications / APIs can provide a more specific meaning, but I would read it to mean "everything about this document is going away".  That may or may not mean the resource itself goes away- we have the `Sunset` header for that anyway.

We are now much more clear about target and context representations and how Hyper-Schema works in general.  If you write a "self" link with a target schema (in `targetSchema`) that is incompatible with or out of sync with your context schema (the one containing the "self" link), then that's your problem.

In general, if there's no sensible interpretation, it can and should be ignored (JSON Schema allows harmless weird things).

> In a property subschema (in any of the property keywords) it means that all matching properties are deprecated but will still be present?

Um... yeah... I'm not sure why I thought this was questionable.

> In a schema applying to a single position in an array/tuple ("items" is an array) it means...?
> In a schema applying to multiple array positions ("items" as one schema, or "additionalItems") it means...?

It's application-specific.  I would guess it means something about the array elements (or certain positions of it) being deprecated without the concept of an array being deprecated.  What that means I'm not sure, but if anyone actually wants to use it presumably they'll provide some guidance on what they mean.  This definitely falls under the "allow harmless apparent nonsense that someone might find a good unanticipated use for anyway" approach.

> There are far too many cases that need discussing and a coherent view of hypermedia resources that need to be developed before we can just toss this in.

...and we have now spent a great deal of effort developing that view.  I'd say it's safe to just toss it in now :-)

Objections?  @awwright ?

**Comment by lashchev** [comment]
@handrews Is there any consensus/progress on this?

**Comment by handrews** [comment]
@philsturgeon want to try to wrap this up for draft-08?

**Comment by philsturgeon** [comment]
This is now part of my sprint. It's got a JIRA issue and everything.

**Comment by philsturgeon** [comment]
Potential implementation over on #737
